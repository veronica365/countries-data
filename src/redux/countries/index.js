import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';
import { API } from '../../constants';

const findItem = (id, countries) => countries.filter((country) => String(country.id) === id);
const searchItem = (param, countries) => countries.filter(
  (country) => String(country.id) === param
      || String(country.name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency).toLowerCase().includes(String(param).toLowerCase())
      || String(country.capital).toLowerCase().includes(String(param).toLowerCase())
      || String(country.region).toLowerCase().includes(String(param).toLowerCase())
      || String(country.currency_name).toLowerCase().includes(String(param).toLowerCase())
      || String(country.subregion).toLowerCase().includes(String(param).toLowerCase()),
);

export const getData = createAsyncThunk('get', async (_, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return countries;
    const { data } = await axios.get(`${API}`);
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const findCountry = createAsyncThunk('find', async (id, thunkAPI) => {
  try {
    const { countries } = thunkAPI.getState().countries;
    if (countries.length) return findItem(id, countries);
    const { data } = await axios.get(`${API}`);
    return findItem(id, data);
  } catch (error) {
    return thunkAPI.rejectWithValue('Something went wrong');
  }
});
export const searchCountry = createAsyncThunk(
  'search',
  async (param, thunkAPI) => {
    try {
      const { countries } = thunkAPI.getState().countries;
      if (countries.length) return searchItem(param, countries);
      const { data } = await axios.get(`${API}`);
      return searchItem(param, data);
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  },
);

const countrieslice = createSlice({
  name: 'countries',
  initialState: {
    countries: [],
    countriesBySearch: [],
    countryById: {},
    isLoading: false,
    error: null,
  },
  reducers: {
    setCountryById: (state, action) => ({
      ...state,
      countryById: action.payload,
      countriesBySearch: [],
      isLoading: false,
    }),
  },
  extraReducers: (builder) => {
    builder.addCase(getData.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(getData.fulfilled, (state, action) => ({
      ...state,
      countries: action.payload,
      isLoading: false,
    }));
    builder.addCase(getData.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(findCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(findCountry.fulfilled, (state, action) => ({
      ...state,
      countryById: action.payload[0] || {},
      isLoading: false,
    }));
    builder.addCase(findCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.pending, (state) => ({
      ...state,
      isLoading: true,
    }));
    builder.addCase(searchCountry.fulfilled, (state, action) => ({
      ...state,
      countriesBySearch: action.payload,
      isLoading: false,
    }));
    builder.addCase(searchCountry.rejected, (state, action) => ({
      ...state,
      error: action.payload,
      isLoading: false,
    }));
  },
});

export const { setCountryById } = countrieslice.actions;

export default countrieslice.reducer;
