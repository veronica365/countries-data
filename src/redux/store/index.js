import { configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import countryReducer from '../countries';

const store = configureStore({
  reducer: {
    countries: countryReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
  devTools: process.env.NODE_ENV !== 'production',
});

export default store;
