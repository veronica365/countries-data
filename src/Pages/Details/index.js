import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import Headline from '../../components/Headline';
import { findCountry, setCountryById } from '../../redux/countries';
import CountryList from '../../components/CountryList';

export default function Details() {
  const { id } = useParams();
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(setCountryById({}));
    dispatch(findCountry(id));
  }, [id, dispatch]);
  return (
    <div className="home">
      <Headline />
      <CountryList id />
    </div>
  );
}
