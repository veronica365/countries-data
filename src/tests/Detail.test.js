import { render } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import axios from 'axios';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import Details from '../Pages/Details';
import store from '../redux/store';
import { findCountry } from '../redux/countries';
import { muchCountryData } from './__Mocks__/countries';

jest.mock('axios');
describe('Details page', () => {
  beforeEach(async () => {
    jest.resetAllMocks();
    jest.resetModules();
    jest.restoreAllMocks();
    axios.get.mockResolvedValue({
      data: [muchCountryData],
    });
  });

  it('should render without issues', async () => {
    await store.dispatch(findCountry(1));
    await act(async () => {
      const route = '/country/1';
      const { container } = render(
        <Provider store={store}>
          <MemoryRouter initialEntries={[route]}>
            <Routes>
              <Route path="/country/:id" element={<Details />} />
            </Routes>
          </MemoryRouter>
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  it('Should be able to fetch data for all countries', async () => {
    const { type, payload } = await store.dispatch(findCountry('1'));
    expect(type).toEqual('find/fulfilled');
    expect(payload[0]).toEqual(muchCountryData);
    expect(store.getState().countries.countryById).toEqual(muchCountryData);
  });
});
