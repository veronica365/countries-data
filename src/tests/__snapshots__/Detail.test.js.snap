// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details page should render without issues 1`] = `<div />`;
